ARM: dts: ux500: samsung-golden: Add touchkey

AffeAli opened the pull request:
The back/menu keys are supposed to be controlled by a Coreriver TC360. I intended to write a driver for it, turns out there is a compatible driver.
`CONFIG_I2C_GPIO` and `CONFIG_KEYBOARD_TM2_TOUCHKEY` need to be enabled.

AffeAli committed:
ARM: dts: ux500: samsung-golden: Add touchkey

stephan-gh commented on arch/arm/boot/dts/ste-ux500-samsung-golden.dts:
```diff
@@ -396,6 +422,25 @@
+	};
+
+	tc360 {
+		tc360_default: tc360_default {
```
I'd call this (and the node above this) `touchkey` instead of `tc360` to make it a bit more generic.

AffeAli replied:
I changed the name.

stephan-gh commented on arch/arm/boot/dts/ste-ux500-samsung-golden.dts:
```diff
@@ -24,6 +24,32 @@
+			compatible = "cypress,midas-touchkey";
+			reg = <0x20>;
+			vdd-supply = <&ab8500_ldo_aux6_reg>;
+			vcc-supply = <&ab8500_ldo_aux4_reg>;
```
I suggest swapping these two regulators. The downstream kernel and schematic I have does not make clear which one of them is which, but the `vdd-supply` is used for LED brightness control in the driver:

https://github.com/novathor-mainline/linux/blob/231702365892778b7f27f9b8d572cce36b94140e/drivers/input/keyboard/tm2-touchkey.c#L86-L95

And note how it changes the voltage to 2.5V / 3.3V to turn the LED on/off. At the moment this causes errors every time you change the brightness:

```
[  141.629001] touch_key_2.2v: voltage operation not allowed
[  148.248833] touch_key_2.2v: voltage operation not allowed
[  152.262619] touch_key_2.2v: voltage operation not allowed
```

Therefore I think it is more appropriate to use the `key_led_3.3v` regulator (`ab8500_ldo_aux4_reg`) as `vdd-supply`.

Now the error still appears when you change the brightness to `0` (but not anymore for `1`):

```
[   50.498276] key_led_3.3v: voltage operation not allowed
[   68.321245] key_led_3.3v: voltage operation not allowed
```

This is because the regulators in `samsung-golden` have a fixed voltage, and should not be changed to another voltage. Fortunately, looking at the code above there is already a `fixed_regulator` option in the driver that prevents changing the regulator voltage.

Therefore I suggest that you:
1. Swap the two regulators as suggested above.
2. Define a new `tc360_touchkey_variant` that looks like `midas_touchkey_variant`,  
but with `.fixed_regulator = true`
3. Use it with a new `coreriver,tc360-touchkey` compatible

Another advantage of the `fixed_regulator` option is that it properly synchronizes the initial state of the touch key backlight. `brightness` is set to `1` after boot, but the backlight is actually off. With `fixed_regulator`, it gets enabled during bootup:

https://github.com/novathor-mainline/linux/blob/231702365892778b7f27f9b8d572cce36b94140e/drivers/input/keyboard/tm2-touchkey.c#L279-L280

Please split this into two commits, one for the input driver change, and one for the dts change.

If you need any help, please feel free to ask here or on Matrix/IRC. I tested this already but this is your patch and I don't want to spoil the fun ;)

stephan-gh reviewed (changes requested):
Thanks again for the great patch!
Indeed, the `tc360` driver downstream looks very similar to this mainline driver.

I have tested your changes on my `samsung-golden` and overall they work great! Touching the keys is recognized correctly. But I have one suggestion regarding the LED backlight handling below:

Minecrell commented:
Thanks! You have also added it to the binding documentation. :+1: 

Preferably binding patches should be in their own commit (see https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/submitting-patches.txt) and come before the commit that make use of the binding in the C code.

Also, for upstreaming it would be helpful if you add the new config options to the `arch/arm/configs/u8500_defconfig`. You can edit it most easily by:
  - Applying it: `make u8500_defconfig`
  - Use your favorite kconfig editor (`menuconfig`, `nconfig`, ...) and make the necessary changes
  - `make savedefconfig`
  - `cp [output]/defconfig arch/arm/configs/u8500_defconfig`
  - Make sure to commit only relevant additions, sometimes options are moved around.
    - Can use e.g. `git add --patch` for this, then it will ask you which hunks you would like to commit.

Note that you cannot use `u8500_defconfig` as-is for postmarketOS. To generate the configuration for `linux-postmarketos-novathor` I use `make u8500_defconfig pmos.config`, which enables some pmos-specific options on top of `u8500_defconfig`.

In total, I would suggest the following 5 commits (in this order):
  1. Add `coreriver` to `Documentation/devicetree/bindings/vendor-prefixes.yaml`
  2. Add `coreriver,tc360-touchkey` to `bindings/input/cypress,tm2-touchkey.txt`
  3. Implement `coreriver,tc360-touchkey` in the driver
  4. Add it to `samsung-golden` device tree.
  5. Enable necessary options in u8500_defconfig.

AffeAli committed:
Input: tm2-touchkey - add support for Coreriver TC360 variant

The Coreriver TouchCore 360 is like the midas board touchkey, but it is using a fixed regulator.

AffeAli committed:
dt-bindings: input: add Coreriver TC360 binding

AffeAli committed:
dt-bindings: vendor-prefixes: Add Coreriver vendor prefix

AffeAli committed:
ARM: defconfig: u8500: Enable CONFIG_KEYBOARD_TM2_TOUCHKEY

Touchkey support on samsung-golden.

Minecrell commented:
Seems to be working fine now, thanks again!

For this repository I try to be very close to mainline - it is mainly a place to put patches that were not applied yet, or need some additional work. In particular, most of the `samsung-golden` patches have already been applied upstream.

Therefore I encourage you to send your patches upstream as well. I would do this in two steps:
  1. The input driver + binding documentation patches to the input mailing list.
  2. Once that is applied the ARM: patches to the arm-kernel mailing list.

`git send-email` and `scripts/get_maintainers.pl` are quite useful, also see [Submitting patches: the essential guide to getting your code into the kernel](https://www.kernel.org/doc/html/latest/process/submitting-patches.html) if you are not familiar with the process.

Please let me know if you have any questions :)
